src/Engine/UnitActionLifecycles/BattleLifecycle.java

```diff
@@ -218,7 +221,7 @@ public Unit getAttacker()
     }
     public boolean attackerDies()
     {
-      return battleInfo.attacker.getPreciseHP() - battleInfo.attackerHPLoss <= 0;
+      return (int) ((battleInfo.attacker.getPreciseHP() - battleInfo.attackerHPLoss) * 10) <= 0;
```

Comment:
You are storing HP as an integer, converting it to a double to pass here, then converting back to an int and truncating. This feels more complicated; does it actually grant better accuracy (or better adherence to cart behavior if you are optimizing for that)? I guess truncating this will technically deliver a different result for values in the range (0, 0.1), but those are all precluded on the `Unit` side.

Reply:
You're right, it is more complicated.
Cart behavior uses exclusively int math, I think, which would require mucking with how the combat math is set up. It's something I'm keeping generally in mind, but not really going for broke on it.

The reason I am converting it to a double at all is because it keeps the `getHP()` and `getPreciseHP()` functions producing the same denomination of health.
I guess it's not out of the question to make a third function for the integer state (or I guess I could just make `health` public), but I'd really like it to be obvious from the accessing syntax that these are not directly-comparable quantities.

The reason I added this cast right here is because I was seeing a few zombie (0 hp) units pop up in testing, so that logic was necessary to sync up with how the unit would actually store the result.

Reply:
> but I'd really like it to be obvious from the accessing syntax that these are not directly-comparable quantities

We could just go all strong-OO and make a `UnitHP` object/datatype with accessors and whatnot, but actually let's not.
Maybe just keep it as the 0-100 percentage value at all times except when displaying it? That's closer to what I expected you meant when you talked about using integer calculations. The tricky bit there is that we'd have to get used to thinking in one scale despite seeing in another.

Reply:
Hmm. You make a good point. Does that sound like it's in scope for this PR?

Reply:
Well, given the name of this branch and your initial comments when discussing the change I thought that was the main point of this PR; at this point I'm not sure what the scope is supposed to be anymore.
So uh, sure.

Reply:
I wanted to do two things, mostly:
Get rid of `BattleInstance`. Every time I went to look at the combat code, I'd click on `CombatEngine` and be bitterly disappointed.
Introduce a split of at least two-fold for dynamic attack modifiers, to support dynamic attack buffs vs non-units.

I also changed the unit HP storage since I figured "hey, we're fiddling with stuff, why not?"... and then I realized it wasn't quite as easy to do as I had thought. (Funny, I had the exact same problem when I was initially drafting the `Unit` class and stuff.)

Reply:
> Every time I went to look at the combat code, I'd click on CombatEngine and be bitterly disappointed

Yeah, I'm pretty sure I did the same. Putting logic where it sounds like it should be is a Good Thing.

> hey, we're fiddling with stuff

Well, you're already fiddling with it; may as well do it proper.